pkg/providers/amifamily/ami.go

```diff
@@ -126,163 +81,128 @@ func (p *DefaultProvider) List(ctx context.Context, nodeClass *v1beta1.EC2NodeCl
 					})
 				} else {
-					elem.Filters = append(elem.Filters, &ec2.Filter{
+					query.Filters = append(query.Filters, &ec2.Filter{
 						Name:   aws.String(fmt.Sprintf("tag:%s", k)),
 						Values: []*string{aws.String(v)},
 					})
 				}
 			}
-			res = append(res, elem)
+			queries = append(queries, query)
 		}
 	}
 	if len(idFilter.Values) > 0 {
-		res = append(res, FiltersAndOwners{Filters: []*ec2.Filter{idFilter}})
+		queries = append(queries, DescribeImageQuery{Filters: []*ec2.Filter{idFilter}})
 	}
-	return res
+	return queries, nil
 }
 
-func (p *DefaultProvider) getRequirementsFromImage(ec2Image *ec2.Image) scheduling.Requirements {
-	requirements := scheduling.NewRequirements()
-	// Always add the architecture of an image as a requirement, irrespective of what's specified in EC2 tags.
-	architecture := *ec2Image.Architecture
-	if value, ok := v1beta1.AWSToKubeArchitectures[architecture]; ok {
-		architecture = value
+func (p *DefaultProvider) amis(ctx context.Context, queries []DescribeImageQuery) (AMIs, error) {
+	hash, err := hashstructure.Hash(queries, hashstructure.FormatV2, &hashstructure.HashOptions{SlicesAsSets: true})
+	if err != nil {
+		return nil, err
 	}
-	requirements.Add(scheduling.NewRequirement(v1.LabelArchStable, v1.NodeSelectorOpIn, architecture))
-	return requirements
+	if images, ok := p.cache.Get(fmt.Sprintf("%d", hash)); ok {
+		// Ensure what's returned from this function is a deep-copy of AMIs so alterations
+		// to the data don't affect the original
+		return append(AMIs{}, images.(AMIs)...), nil
+	}
+	images := map[uint64]AMI{}
+	for _, query := range queries {
+		if err = p.ec2api.DescribeImagesPagesWithContext(ctx, query.DescribeImagesInput(), func(page *ec2.DescribeImagesOutput, _ bool) bool {
+			for _, image := range page.Images {
+				arch, ok := v1beta1.AWSToKubeArchitectures[lo.FromPtr(image.Architecture)]
+				if !ok {
+					continue
+				}
+				// Each image may have multiple associated sets of requirements. For example, an image may be compatible with Neuron instances
+				// and GPU instances. In that case, we'll have a set of requirements for each, and will create one "image" for each.
+				for _, reqs := range query.RequirementsForImageWithArchitecture(lo.FromPtr(image.ImageId), arch) {
+					// If we already have an image with the same set of requirements, but this image is newer, replace the previous image.
+					reqsHash := lo.Must(hashstructure.Hash(reqs.NodeSelectorRequirements(), hashstructure.FormatV2, &hashstructure.HashOptions{SlicesAsSets: true}))
+					if v, ok := images[reqsHash]; ok {
+						candidateCreationTime, _ := time.Parse(time.RFC3339, lo.FromPtr(image.CreationDate))
+						existingCreationTime, _ := time.Parse(time.RFC3339, v.CreationDate)
+						if existingCreationTime == candidateCreationTime && lo.FromPtr(image.Name) < v.Name {
+							continue
+						}
+						if candidateCreationTime.Unix() < existingCreationTime.Unix() {
+							continue
+						}
```

Comment:
Seems like the check should be 
1. Check if time A is less than time B
2. If times are equal, check if name A < name B


I think this is the same, but it confuses because we check the higher priority check second.

Reply:
synced offline, going to keep as is